File: lc_p75.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int size = nums.size();
        int red = 0;
        int blue = size - 1;
        int i = 0;

        while (i <= blue) {
            if (nums[i] == 0) { // 红色
                std::swap(nums[i++], nums[red++]);
            } else if (nums[i] == 1) { // 白色
                i++;
            } else if (nums[i] == 2) { // 蓝色
                std::swap(nums[i], nums[blue--]);
            } else {
                break;
            }
        }
    }
};

int main(void) {
    std::vector<int> nums = {2, 0 , 1};
    Solution().sortColors(nums);

    for (const auto& i : nums) {
        std::cout << i << " ";
    }
    
    return 0;
}

File: lc_p75.py
from typing import *

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        red: int = 0
        blue: int = len(nums) - 1
        i = 0

        while i <= blue:
            if nums[i] == 0:
                nums[i], nums[red] = nums[red], nums[i]
                i += 1
                red += 1
            elif nums[i] == 1:
                i += 1
            elif nums[i] == 2:
                nums[i], nums[blue] = nums[blue], nums[i]
                blue -= 1
            else:
                break


def main() -> None:
    nums = [2, 0, 1]
    Solution().sortColors(nums)
    print(nums)

if __name__ == "__main__":
    main()
